File: generate_sitemap.py
#!/usr/bin/env python3
"""
Generates a dynamic sitemap.xml from articles.js
Run this whenever you add new articles to automatically update the sitemap
"""

import re
from datetime import datetime
from pathlib import Path

def extract_categories_from_js():
    """Extract category names from articles.js"""
    articles_file = Path(__file__).parent / "src" / "articles.js"
    
    with open(articles_file, 'r', encoding='utf-8') as f:
        content = f.read()
    
    # Extract category names using regex
    categories = re.findall(r'"([^"]+)":\s*\[', content)
    return list(dict.fromkeys(categories))  # Remove duplicates, keep order

def generate_sitemap(categories, domain="https://grüner-faktencheck.de"):
    """Generate sitemap XML content"""
    sitemap_lines = [
        '<?xml version="1.0" encoding="UTF-8"?>',
        '<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">',
        '  <url>',
        f'    <loc>{domain}/</loc>',
        f'    <lastmod>{datetime.now().strftime("%Y-%m-%d")}</lastmod>',
        '    <changefreq>weekly</changefreq>',
        '    <priority>1.0</priority>',
        '  </url>',
    ]
    
    # Add category pages
    for category in categories:
        category_slug = category.lower().replace(" ", "-")
        category_url = f"{domain}/category/{category_slug}"
        sitemap_lines.append('  <url>')
        sitemap_lines.append(f'    <loc>{category_url}</loc>')
        sitemap_lines.append(f'    <lastmod>{datetime.now().strftime("%Y-%m-%d")}</lastmod>')
        sitemap_lines.append('    <changefreq>weekly</changefreq>')
        sitemap_lines.append('    <priority>0.8</priority>')
        sitemap_lines.append('  </url>')
    
    sitemap_lines.append('</urlset>')
    
    return '\n'.join(sitemap_lines)

if __name__ == "__main__":
    try:
        categories = extract_categories_from_js()
        sitemap_content = generate_sitemap(categories)
        
        sitemap_path = Path(__file__).parent / "public" / "sitemap.xml"
        with open(sitemap_path, 'w', encoding='utf-8') as f:
            f.write(sitemap_content)
        
        print(f"✓ Sitemap generated: {sitemap_path}")
        print(f"✓ Categories added: {', '.join(categories)}")
    except Exception as e:
        print(f"✗ Error: {e}")
        exit(1)

File: scripts/generate_sitemap.js
const fs = require('fs');
const path = require('path');
const vm = require('vm');

const SRC = path.join(__dirname, '..', 'src', 'articles-enhanced.js');
const OUT = path.join(__dirname, '..', 'public', 'sitemap.xml');
const DOMAIN = 'https://gruener-faktencheck.de';

function loadArticles() {
  const src = fs.readFileSync(SRC, 'utf8');
  // Remove "export" and wrap in an object for evaluation
  const transformed = src.replace(/^\s*export\s+const\s+articles\s*=\s*/m, 'const articles = ');
  // Execute safely in VM
  const sandbox = {};
  vm.createContext(sandbox);
  try {
    vm.runInContext(transformed + '\n;globalThis.__ARTICLES__ = articles;', sandbox, { timeout: 1000 });
    return sandbox.__ARTICLES__;
  } catch (err) {
    console.error('Failed to parse articles file:', err.message);
    process.exit(1);
  }
}

function buildUrls(articles) {
  const urls = new Set();
  // static pages
  ['/', '/aussenpolitik.html', '/energie.html', '/innenpolitik.html', '/wirtschaft.html'].forEach(p => urls.add(DOMAIN + p));

  // categories
  Object.keys(articles).forEach(cat => {
    const slug = cat.toLowerCase().replace(/ä/g,'ae').replace(/ö/g,'oe').replace(/ü/g,'ue').replace(/ß/g,'ss').replace(/[^a-z0-9]+/g,'-').replace(/^-|-$/g,'');
    urls.add(`${DOMAIN}/category/${slug}`);
    // article links (external) are included as outbound sources, but do not include external links in sitemap
  });

  return Array.from(urls);
}

function writeSitemap(list) {
  const lines = [];
  lines.push('<?xml version="1.0" encoding="UTF-8"?>');
  lines.push('<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">');
  const now = new Date().toISOString().split('T')[0];
  list.forEach(loc => {
    lines.push('  <url>');
    lines.push(`    <loc>${loc}</loc>`);
    lines.push(`    <lastmod>${now}</lastmod>`);
    lines.push('    <changefreq>weekly</changefreq>');
    lines.push('    <priority>0.8</priority>');
    lines.push('  </url>');
  });
  lines.push('</urlset>');

  fs.writeFileSync(OUT, lines.join('\n'));
  console.log('Wrote sitemap to', OUT);
}

function main() {
  const articles = loadArticles();
  const urls = buildUrls(articles);
  writeSitemap(urls);
}

if (require.main === module) main();
